File: NeetCode 150/1-D DP/Coin Change/322.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
  int minimum(int a, int b) {
    if (a!=-1 && b!=-1) return min(a,b);
    else return max(a,b);
  }

  // Normal recursion
  int recursion(int count, vector<int>& coins, int curCoin, int remainder) {
    // base case
    if(remainder == 0) return count;
    if(remainder < 0) return -1;
    if(curCoin < 0) return -1;

    int a = recursion(count+1, coins, curCoin, remainder-coins[curCoin]); // - biggest
    int b = recursion(count, coins, curCoin-1, remainder); // forget about biggest coin
    return minimum(a,b);
  }

  // DP
  int coinChange(vector<int>& coins, int amount) {
    sort(coins.begin(), coins.end());
    // index == amount, dp[index] = min coins to reach amount
    vector<int> dp = vector<int>(amount+1, INT_MAX);
    dp[0] = 0;
    for(int i = 1; i < dp.size(); i++) {
      for(int c : coins) {
        if(i == c) { dp[i] = 1; break; }
        if(i - c < 0) break;
        if(dp[i-c] != INT_MAX) dp[i] = min(dp[i], dp[i-c]+1);
      }
    }
    return dp[amount] == INT_MAX ? -1 : dp[amount];
  }
};

File: NeetCode 150/1-D DP/Coin Change/322.py
class Solution(object):
    # def recurse(self, count, coins, remainder):
    #     if remainder == 0:
    #         return count
    #     if remainder < 0:
    #         return -1
    #     if (len(coins) == 0):
    #         return -1
    #     with_largest_coin = self.recurse(count+1, coins, remainder-coins[len(coins)-1])
    #     without_largest_coin = self.recurse(count, coins[:-1], remainder)

    #     if with_largest_coin != -1 and without_largest_coin != -1:
    #         return min(with_largest_coin,without_largest_coin)
    #     elif (with_largest_coin != -1):
    #         return with_largest_coin
    #     else:
    #         return without_largest_coin

    def coinChange(self, coins, amount):
      """
      :type coins: List[int]
      :type amount: int
      :rtype: int
      """
      solution, found = self.myCoinChange(self, coins, amount, False)
      if found:
        return solution
      else:
        return -1
        
    def myCoinChange(self, coins, amount, found):
      if amount == 0:
        found = True
        return 1, found
      elif amount < 0 or coins.size() == 0:
        return -1, found
      else:
        return (self.myCoinChange(coins, amount-coins(coins.size()-1), found) + self.myCoinChange(coins[:-1], amount, found)), found
        

        

    

    
        